### PR TITLE
[RyuJIT/X86] Limit byte register candidates for CRC32

### LINKED ISSUE #14771: [RyuJIT/x86][JitStressRegs=0x80] Assertion failed 'emitVerifyEncodable(ins, size, reg1, reg2)'
This assert occurs in this recently-added test:
JIT\HardwareIntrinsics\Crc32_r\Crc32_r.exe

This impedes arm altjit stress testing.

### PR DESCRIPTION
**Pull request by fiigii**
Fix #14771

### CODE REVIEW

**Comment by fiigii** [comment]
@dotnet-bot test Windows_NT jitstressregs0x80

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT x86 Checked jitstressregs0x80

File: src/jit/lsraxarch.cpp [hunk 1]
```diff
@@ -2437,6 +2437,21 @@ void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
         info->srcCount += GetOperandSourceCount(intrinsicTree->gtOp.gtOp2);
     }
     info->srcCount += GetOperandSourceCount(intrinsicTree->gtOp.gtOp1);
+
+#ifdef _TARGET_X86_
+    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
+    {
+        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS could be used as byte registers.
```
Comment (BruceForstall) [inline src/jit/lsraxarch.cpp]: grammar: could => can

File: src/jit/lsraxarch.cpp [hunk 2]
```diff
@@ -2437,6 +2437,21 @@ void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
...
+        //
+        // TODO - currently we use the BaseType to bring the type of the secnod argument
```
Comment (BruceForstall) [inline src/jit/lsraxarch.cpp]: typo: secnod => second

File: src/jit/lsraxarch.cpp [hunk 3]
```diff
@@ -2437,6 +2437,21 @@ void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
         info->srcCount += GetOperandSourceCount(intrinsicTree->gtOp.gtOp2);
     }
     info->srcCount += GetOperandSourceCount(intrinsicTree->gtOp.gtOp1);
+
+#ifdef _TARGET_X86_
+    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
```
Comment (BruceForstall) [inline src/jit/lsraxarch.cpp]: Is this specific to CRC32 op2? Do we need a more general mechanism here to specify if op1 or op2 required byte registers on x86?

File: src/jit/lsraxarch.cpp [hunk 4]
```diff
@@ -2437,6 +2437,21 @@ void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
...
+    {
+        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS could be used as byte registers.
+        //
+        // TODO - currently we use the BaseType to bring the type of the secnod argument
+        // to the code generator. May encode the overload info in other way.
+        var_types srcType = intrinsicTree->gtSIMDBaseType;
+        if (srcType == TYP_UBYTE || srcType == TYP_BYTE)
```
Comment (BruceForstall) [inline src/jit/lsraxarch.cpp]: Should you use `varTypeIsByte()` to also include TYP_BOOL?

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib

File: src/jit/lsraxarch.cpp (hunk 3 above)
Comment (fiigii) [reply src/jit/lsraxarch.cpp]: > Is this specific to CRC32 op2? Do we need a more general mechanism here to specify if op1 or op2 required byte registers on x86?

I believe that only CRC32 will use "byte".

File: src/jit/lsraxarch.cpp (hunk 1 above)
Comment (fiigii) [reply src/jit/lsraxarch.cpp]: Thanks!

File: src/jit/lsraxarch.cpp (hunk 2 above)
Comment (fiigii) [reply src/jit/lsraxarch.cpp]: Will fix!

File: src/jit/lsraxarch.cpp (hunk 4 above)
Comment (fiigii) [reply src/jit/lsraxarch.cpp]: Good point! Will do.

**Comment by fiigii** [comment]
@dotnet-bot test Windows_NT x86 Checked jitstressregs0x80

**Comment by fiigii** [comment]
@dotnet-bot test Windows_NT x86 Checked jitstressregs0x80

**Comment by fiigii** [comment]
@dotnet-bot test Windows_NT x86 Checked jitstressregs0x80

**Comment by BruceForstall** [comment]
@dotnet-bot
test Windows_NT x86 Checked jitstressregs0x80
test CentOS7.1 x64 Checked Innerloop Build and Test
test OSX10.12 x64 Checked Innerloop Build and Test
test Ubuntu x64 Checked Innerloop Build and Test

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu x64 Innerloop Formatting

**Pull request merged**